lib/forcegraph.js

```diff
@@ -1,744 +1,226 @@
-        return distanceLink(e, d.source, d.target) < LINE_RADIUS;
-      });
+      e = {x: e[0], y: e[1]};
 
-      if (links.length > 0) {
-        router.link(links[0].o)();
-      }
-    }
 
-    function zoom(z, scale) {
-      var size = getSize();
-      var newSize = [size[0] / scale, size[1] / scale];
-
-      var sidebarWidth = sidebar();
-      var delta = [size[0] - newSize[0], size[1] - newSize[1]];
-      var translate = z.translate();
-      var translateNew = [sidebarWidth + (translate[0] - sidebarWidth - delta[0] / 2) * scale, (translate[1] - delta[1] / 2) * scale];
+      var closedLink;
+      var radius = LINK_RADIUS_SELECT;
+      intLinks
+        /* Disable Clickable VPN
+        .filter(function (d) {
+          return d.o.type !== 'fastd' && d.o.type !== 'L2TP';
+        })
+        */
+        .forEach(function (d) {
+          var distance = math.distanceLink(e, d.source, d.target);
+          if (distance < radius) {
+            closedLink = d;
+            radius = distance;
+          }
+        });
 
-      animatePanzoom(translateNew, z.scale() * scale);
+      if (closedLink !== undefined) {
+        router.link(closedLink.o)();
+      }
     }
 
-    function keyboardZoom(z) {
-      return function () {
-        var e = d3.event;
-
-        if (e.altKey || e.ctrlKey || e.metaKey) {
-          return;
-        }
+    function redraw() {
+      ctx.save();
+      ctx.clearRect(0, 0, canvas.width, canvas.height);
+      ctx.translate(transform.x, transform.y);
+      ctx.scale(transform.k, transform.k);
 
-        if (e.keyCode === 43) {
-          zoom(z, 1.41);
-        }
+      intLinks.forEach(draw.drawLink);
+      intNodes.forEach(draw.drawNode);
 
-        if (e.keyCode === 45) {
-          zoom(z, 1 / 1.41);
-        }
-      };
+      ctx.restore();
     }
 
     el = document.createElement('div');
     el.classList.add('graph');
 
-    font = window.getComputedStyle(el).fontSize + ' ' + window.getComputedStyle(el).fontFamily;
-
-    zoomBehavior = d3.behavior.zoom()
-      .scaleExtent([1 / 3, 3])
-      .on('zoom', onPanZoom)
-      .translate([sidebar(), 0]);
+    forceLink = d3Force.forceLink()
+     .distance(function (d) {
+       if (d.o.type === 'fastd' || d.o.type === 'L2TP') {
+         return 0;
+       }
+       return 75;
+     })
+     .strength(function (d) {
+       if (d.o.type === 'fastd' || d.o.type === 'L2TP') {
+         return 0.02;
+       }
+       return Math.max(0.5, 1 / d.o.tq);
+     });
+
+    var zoom = d3Zoom.zoom()
+         .scaleExtent([ZOOM_MIN, ZOOM_MAX])
+         .on('zoom', function () {
+           transform = d3Selection.event.transform;
+           draw.setTransform(transform);
+           redraw();
+         });
+
+
+    force = d3Force.forceSimulation()
+      .force('link', forceLink)
+      .force('charge', d3Force.forceManyBody())
+      .on('tick', redraw)
+      .alphaTarget(0.1);
+
+    var drag = d3Drag.drag()
+      .subject(function () {
```

Comment:
allgemein passt das nicht, glaube function is nur ein space, paar doppelte leerzeilen